File: web/src/core/tfl/format.ts
// TFL Document formatter — ported from Organon.Syl.Lsp.Format (adapted for TFL)

/**
 * Format a .tfl document: normalize whitespace, canonicalize conclusion
 * markers, collapse blank lines.
 */
export function formatText(text: string): string {
  let lines = text.split("\n");
  lines = lines.map(stripEnd);
  lines = lines.map(canonicalizeTherefore);
  lines = normalizeTflLines(lines);
  lines = collapseBlankLines(lines);
  return lines.join("\n");
}

function stripEnd(line: string): string {
  return line.replace(/\s+$/, "");
}

/** Replace a leading "therefore" with "∴" on conclusion lines. */
function canonicalizeTherefore(line: string): string {
  const match = line.match(/^(\s*)therefore\s/i);
  if (match) {
    return match[1] + "∴ " + line.slice(match[0].length);
  }
  return line;
}

/** Normalize indentation and internal whitespace on TFL statement lines. */
function normalizeTflLines(lines: string[]): string[] {
  return lines.map((line) => {
    if (isTflStatementLine(line)) {
      const collapsed = line.trim().replace(/\s+/g, " ");
      return "  " + collapsed;
    }
    return line;
  });
}

function isTflStatementLine(line: string): boolean {
  const stripped = line.replace(/^\s+/, "");
  if (stripped.length === 0) return false;
  // Must have leading whitespace (indented under a proof block)
  if (stripped.length === line.length) return false;
  // TFL statements start with +, -, *, ∴, or English quantifiers
  return (
    stripped.startsWith("+") ||
    stripped.startsWith("-") ||
    stripped.startsWith("*") ||
    stripped.startsWith("∴") ||
    stripped.startsWith("?") ||
    stripped.startsWith("@") ||
    stripped.toLowerCase().startsWith("every ") ||
    stripped.toLowerCase().startsWith("no ") ||
    stripped.toLowerCase().startsWith("some ")
  );
}

/** Collapse runs of 3+ blank lines into 2. */
function collapseBlankLines(lines: string[]): string[] {
  const result: string[] = [];
  let blanks = 0;
  for (const line of lines) {
    if (line.trim() === "") {
      blanks++;
      if (blanks <= 2) result.push(line);
    } else {
      blanks = 0;
      result.push(line);
    }
  }
  return result;
}

File: src/Organon/Tfl/Format.hs
module Organon.Tfl.Format
  ( formatText,
  )
where

import Data.List (groupBy)
import qualified Data.Text as T

-- | Text-level formatting for TFL documents.
-- Strips trailing whitespace, canonicalizes therefore, normalizes
-- indentation on statement lines and collapses blank runs.
formatText :: T.Text -> T.Text
formatText txt =
  let lns = T.lines txt
      trimmed = map T.stripEnd lns
      canonicalized = map canonicalizeTherefore trimmed
      normalized = normalizeTflLines canonicalized
      collapsed = collapseBlankLines normalized
   in T.unlines collapsed

-- | Replace a leading "therefore" with "∴" on conclusion lines.
canonicalizeTherefore :: T.Text -> T.Text
canonicalizeTherefore line =
  let (indent, rest) = T.span (== ' ') line
      lower = T.toLower rest
   in if "therefore " `T.isPrefixOf` lower
        then indent <> "∴ " <> T.drop 10 rest
        else line

-- | Normalize indentation on TFL statement lines.
normalizeTflLines :: [T.Text] -> [T.Text]
normalizeTflLines = map normalizeLine
  where
    normalizeLine line
      | isTflStatementLine line = "  " <> (T.unwords . T.words . T.strip) line
      | otherwise = line
    isTflStatementLine line =
      let stripped = T.stripStart line
       in not (T.null stripped)
            && T.length stripped < T.length line
            && any
              (`T.isPrefixOf` T.toLower stripped)
              [ "+", "-", "−", "*", "∴ ", "?", "@",
                "every ", "no ", "some "
              ]

-- | Collapse consecutive blank lines into a single blank line.
collapseBlankLines :: [T.Text] -> [T.Text]
collapseBlankLines = concatMap squash . groupBy sameBlank
  where
    sameBlank a b = T.null a && T.null b
    squash [] = []
    squash grp@(g : _)
      | T.null g = [T.empty]
      | otherwise = grp
